students/RimleeT/lesson09/mailroom_oo/donor_models.py

```diff
@@ -0,0 +1,81 @@
+import os
+import tempfile
+
+class Donor():
+    def __init__(self, name, donation=None):
+
+        if type(name) != str:
+            raise Exception("Donor name should be a string type")
+        else:
+            self.name = name 
+        
+        self.donations = list()
+        
+        if donation != None:
+            self.donate(donation)
+
+    def donate(self, donation):
+        if type(donation) == list:
+            for item in donation:
+                self.donate(item)
+        else:
+            try:
+                donation_amount = float(donation) # Throws exection for invalid exception
+            except Exception:
+                raise Exception("Invalid type for donation amount")
+            self.donations.append(donation_amount)
+        
+    def get_stat(self):
+        total_donation = float(format(sum(self.donations),'.2f'))
+        num_donation = len(self.donations)
+        avg_donation = float(format(total_donation/num_donation,'.2f'))
+        
+        return {
+                    "Total": total_donation,
+                    "Count": num_donation,
+                    "Avg": avg_donation
+               }
+    
+    def send_letter(self, dir_name):
+        stat = self.get_stat()
+        letter = f"Dear {self.name}\n\n\tThank you for your very kind donation of ${stat['Total']} It will be put to very good use.\n\nSincerely,\nThe Team"
+        
+        f = tempfile.NamedTemporaryFile(prefix = self.name + '-', dir = dir_name, delete = False)
+        f.write(letter.encode())
+        
+        return letter
+        
+class DonorCollection():
+    def __init__(self):
+        self._donor_list = list()
```

Comment:
It may be a bit better to use a Dict instead of an array because any time you want to return a Donor's information, you have to iterate through the list to find the Donor. Instead, if you were to use a Dict, you could map it to the name (since they should be unique) and it would be more performant!